### PR TITLE
Improve anonymous identifiers #5230

### LINKED ISSUE #5230: Improve anonymous identifiers

### PR DESCRIPTION
**Pull request by kanghj**
Fixes #5230. @wkurniawan07, will need your help for the tests

### CODE REVIEW

**Comment by wkurniawan07** [comment]
What exactly is the problem, though? Something to do with sorting?

**Comment by kanghj** [comment]
Yes, we sort the responses/panels in some places by giver/recipient, so if the recipient is now generated based on the encryption key in build.properties, different keys are going to produce different values for the giver/recipients, resulting in a different sort order of the panels. In the diff of this pr, we can already see some places getting swapped

**Comment by wkurniawan07** [comment]
Not sure if this is feasible, but one thing I can think of is keeping two field names, e.g the one previously used, and a new one used to solve this issue. That is, we sort based on the one previously used, display the new one. One down side is the sorting order might give some clue on the identity, but it's still better than when the supposedly anonymous identity is outright obvious.

**Comment by wkurniawan07** [comment]
I just thought of a better alternative. See if we can customize the comparator like this:
- All the anonymous responses to be displayed separately, it doesn't matter if it's first or last.
- Within the anonymous ones, sort the responses by something else predictable like response metadata or something like that.
That way, we can get a fixed order and yet still obtain a much higher anonymity than before.

**Comment by kanghj** [comment]
Sounds good. I can do that as a separate issue first.

**Comment by wkurniawan07** [comment]
Ok then, changing the label.

**Comment by wkurniawan07** [comment]
@kanghj are the tests here stable now?

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 1]
```diff
@@ -387,6 +388,15 @@ private static String replaceUnpredictableValuesWithPlaceholders(String content)
...
                                   + " type=\"hidden\" value=\"\\${regkey\\.enc}\"")
+                      // anonymous student identifier on results page
+                      .replaceAll("Anonymous student " + REGEX_ANONYMOUS_PARTICIPANT_HASH,
+                                  "Anonymous student \\${student\\.hash}")
+                      // anonymous instructor identifier on results page
+                      .replaceAll("Anonymous instructor " + REGEX_ANONYMOUS_PARTICIPANT_HASH,
+                                  "Anonymous instructor \\${instructor\\.hash}")
+                      // anonymous team identifier on results page
+                      .replaceAll("Anonymous team " + REGEX_ANONYMOUS_PARTICIPANT_HASH,
+                                  "Anonymous team \\${team\\.hash}")
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: If the hashes are all the same, I don't think we need three separate replaces. `.replaceAll("Anonymous (student|instructor|team) " + REGEX, "Anonymous $1 \\${participant\\.hash}")` should do the trick.

File: src/test/resources/pages/godmode.html [hunk 2]
```diff
@@ -31,6 +31,9 @@
                 <li><input type="hidden" value="encoded_gs_key:ag_10ZWFtbWF0ZXMtdGh5ch0LEhBGZWVkYm-Fja1F1ZXN0aW9uGICAgICA0NQLDA" name="blob-key" id="blobKey"></li>
                 <li>key=ag_10ZWFtbWF0ZXMtdGh5ch0LEhBGZWVkYm-Fja1F1ZXN0aW9uGICAgICA0NQLDA</li>
                 <li><input type="hidden" value="ag_10ZWFtbWF0ZXMtdGh5ch0LEhBGZWVkYm-Fja1F1ZXN0aW9uGICAgICA0NQLDA" name="key"></li>
+                <li>Anonymous student 1224044052</li>
+                <li>Anonymous team 1224044052</li>
+                <li>Anonymous instructor 1224044052</li>
```
Comment (wkurniawan07) [inline src/test/resources/pages/godmode.html]: Having said that before, we still need all three of these :+1:

**Comment by kanghj** [comment]
@wkurniawan07, ran it twice with different encryption.key values, and both passed. I just noticed a strange part of the code though.

**Comment by wkurniawan07** [comment]
Two things that may need to be done:
- The same procedure may need to be replicated in `FeedbackResponseCommentSearchResultBundle`.
- `Math.abs(name.hashCode())` is a bad idea when the hash code is `Integer.MIN_VALUE`. Not a big issue since the probability is 1 / 2^32, but try to fix it if possible.

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java [hunk 3]
```diff
@@ -1535,6 +1562,12 @@ public void testGetFeedbackSessionResultsSummaryAsCsv() throws Exception {
         session = newDataBundle.feedbackSessions.get("contribSessionStudentAnonymised");
         instructor = newDataBundle.instructors.get("instructor1OfCourse1");
         
+        String student1Name = getAnonName(newDataBundle, "student1InCourse1");
+        String student2Name = getAnonName(newDataBundle, "student2InCourse1");
+        String student3Name = getAnonName(newDataBundle, "student3InCourse1");
+        String student4Name = getAnonName(newDataBundle, "student4InCourse1");
+        String student5Name = getAnonName(newDataBundle, "student5InCourse1");
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: `student*AnonName` would be a better var name.

File: src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java [hunk 4]
```diff
@@ -1726,6 +1759,16 @@ public void testGetFeedbackSessionResultsSummaryAsCsv() throws Exception {
...
+    
+    private String getAnonEmail(DataBundle dataBundle, String studentKey) {
+        return FeedbackSessionResultsBundle.getAnonEmail(FeedbackParticipantType.STUDENTS,
+                                                         dataBundle.students.get(studentKey).name);
+    }
+
+    private String getAnonName(DataBundle dataBundle, String studentKey) {
+        return FeedbackSessionResultsBundle.getAnonName(FeedbackParticipantType.STUDENTS,
+                                                        dataBundle.students.get(studentKey).name);
+    }
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java]: `getStudentAnonEmail|Name` is more representative.

**Comment by wkurniawan07** [comment]
If there are nothing else to add, those two minor fixes should be the last ones to do.

**Pull request merged**